### ISSUE TITLE
bigquery iam permissions not updatable

### ISSUE THREAD

**Issue by tommed** [GoogleCloudPlatform/deploymentmanager-samples#502]
I have an existing environment which I've created and managed via Deployment Manager, including a BigQuery dataset and several tables. 

A requirement came up where I needed a new service account which needed permissions on the BigQuery dataset, so I added the service account and adjusted the dataset's `accessControl` policy in my jinja template.

```yaml
resources:
#...
- name: my_dataset
  type: bigquery.v2.dataset
  accessControl:
    gcpIamPolicy:
      bindings:
      - role: 'roles/bigquery.dataOwner'
        members:
        - "serviceAccount:polyglot-sync@{{ properties["project"] }}.iam.gserviceaccount.com"
      - role: 'roles/bigquery.jobUser'
        members:
        - "serviceAccount:polyglot-sync@{{ properties["project"] }}.iam.gserviceaccount.com"
#...
```

The service account created successfully, but I received an error when DM tried to deploy the new accessControl policy:

```
- code: NO_METHOD_TO_UPDATE_ACCESS_CONTROL
  location: /deployments/sre-staging-17/resources/my_dataset
  message: No method found to update access control on resource 'my_dataset' of type
    'bigquery.v2.dataset'.
```

Surely accessControl policies will always be subject to change? Having them non-updatable seems unbelievable. Have I just misunderstood something, or is there another way to achieve this please?

**Comment by ocsig** [comment]
I've tested your scenario with the `gcp-types/bigquery-v2:datasets` type and update or permissions works fine.

I would also highlight the Cloud Foundation Toolkit templates for Bigquery Dataset and Table:
https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/tree/master/dm/templates/bigquery

Please let me know if this solved your issue!

**Comment by tommed** [comment]
Thanks will try this shortly. Could I ask, what is the difference between using `bigquery.v2.dataset` listed on the [DM support resource types](https://cloud.google.com/deployment-manager/docs/configuration/supported-resource-types) and `gcp-types/bigquery-v2:datasets` listed on your example please?

**Comment by ocsig** [comment]
TL;DR: Use gcp-types where it is possible.

We understand your confusion in types. They are all calling the same API endpoints, but may implement some edge cases differently.

- (Legacy) resource types: Initial implementation of the connections between DM and GCP APIs
- GCP Types: These types are directly using the APIs discovery document and only implement corrections where it is needed. ( For example where an API's update command is named differently or the API is using an unconventional structure which is not part of the automatic discovery.)
- Custom Types: It is the same as GCP Types, but you as a DM user can define your own type. With the help of custom types you can use Alpha, or even invitation only Alpha APIs.

The Deployment Manager team is working on to improve the documentation of these types.

**Comment by tommed** [comment]
Not sure I totally understand, but look forward to seeing some updated documentation for Deployment Manager - if you follow the links in what's live now, the information doesn't match which I why I think this has become confusing. Appreciate you explaining it though, thank you! :)

Back to my blocker, I have updated like so:

```yaml
- name: my_dataset
  type: gcp-types/bigquery-v2:datasets
  properties:
    location: europe-west2
    datasetReference:
      datasetId: my_dataset
    labels:
      deployment_slot: {{ properties["deployment_slot"] }}
    access:
    - role: 'roles/bigquery.dataOwner'
      userByEmail: my-bqsync-user@{{ properties["project"] }}.iam.gserviceaccount.com
    - role: 'roles/bigquery.jobUser'
      userByEmail: my-bqsync-user@{{ properties["project"] }}.iam.gserviceaccount.com
```

and now get the error:

```
- code: RESOURCE_ERROR
  location: /deployments/sre-staging-20/resources/my_dataset
  message: '{"ResourceType":"gcp-types/bigquery-v2:datasets","ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"errors":[{"domain":"global","message":"Email
    address contains invalid domain name","reason":"invalid"}],"message":"Email address
    contains invalid domain name","statusMessage":"Bad Request","requestPath":"https://bigquery.googleapis.com/bigquery/v2/projects/my_project-staging/datasets","httpMethod":"POST"}}'
```

**Comment by ocsig** [comment]
Can you please verify the Expanded Config of the deployment and confirm the userByEmail value is the correct SA IAM? If yes, please let me know because it smells like a bug.

As a workaround can you try this:
`iamMember:  serviceAccount:my-bqsync-user@{{ properties["project"] }}.iam.gserviceaccount.com`
instead of `userByEmail`?

**Comment by ocsig** [comment]
The reaseon behind the lack of verification at iamMember is that it's possible to add permissions for entities outside your organization, where you can't verify the existance of the user.

About your issue with attaching permissions:
Are you still attaching these permissions to the dataset? Based on the [API docs](https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets), I believe it only accepts 3 roles: [OWNER, READER, WRITER]. Not sure why `roles/bigquery.dataOwner` is not added but `roles/bigquery.jobUser` looks invalid on the dataset. ( Not a BQ IAM expert.)

I can say, the Cloud Foundation Toolkit template works for me with this conig as well:

```
imports:
  - path: ../templates/bigquery/bigquery_dataset.py
    name: bigquery_dataset.py

resources:
  - name: test-bq-dataset
    type: bigquery_dataset.py
    properties:
      name: test_bq_dataset
      location: US
      access:
        - role: OWNER
          userByEmail: demo@cft.tips
        - role: OWNER
          userByEmail: cft-demo-sa-0@acc-cft-dev-1.iam.gserviceaccount.com
        - role: READER
          userByEmail: forseti-server-gcp-11111@forseti-111111.iam.gserviceaccount.com
```

**Comment by tommed** [comment]
Thanks, the lack of verification insight is useful to know. 

The docs do say:

> The following legacy mappings will be applied: OWNER <=> roles/bigquery.dataOwner WRITER <=> roles/bigquery.dataEditor READER <=> roles/bigquery.dataViewer **This field will accept any of the above formats, but will return only the legacy format**. For example, if you set this field to "roles/bigquery.dataOwner", it will be returned back as "OWNER".

I interpreted this to mean you can pass in valid roles such as `roles/bigquery.jobUser`, but just don't expect it to return this in the same output. I am puzzled why the IAM here would only support what is referred to as "legacy" permissions and not actual role names; but assume this will change as DM evolves?

I made a change to reduce my access entry to just:

```yaml
access:
  - role: WRITER
    userByEmail: iamMember: 'serviceAccount:my-bqsync-user@{{ env["project"] }}.iam.gserviceaccount.com'
```
However `bq show` still shows only `projectReaders`, `projectWriters` and `projectOwners`, suggesting it still hasn't assigned the permissions correctly.

I do appreciate you saying the python examples work, but would really like to get the vanilla yaml working here before adding further complexity into our approach.
